### ISSUE TITLE
Add a `package-lock.json`

### ISSUE THREAD

**Issue by gekoke** [mdx-js/mdx-analyzer#336]
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

The repository does not include a `package-lock.json` file. This makes it difficult to build the project reproducibly. [Nixpkgs](https://github.com/NixOS/nixpkgs) specifically relies on npm projects having lockfiles to package them.

I can see that this is a conscious decision, as noted in https://github.com/mdx-js/mdx-analyzer/commit/544560843e4282bcb37b78563b7d61eb44de8d44. Is there a particular reason why things are this way?

### Solution

Remove `package-lock = false` form `.npmrc` and run `npm i`

### Alternatives

Downstream packagers could maintain their own version of the lockfile. This would be undesirable from their perspective.

**Comment by ChristianMurphy** [comment]
Welcome @gekoke! 👋 

TL;DR we don't publish to application registries (`aptitute`, `yum`, `snapcraft`, `nixpkgs`, etc), and have no current plans to.

I'm a bit unclear why we would publish to `nixpkgs` or any other application registry.
I'd view the code here mostly as libraries to build an application (not an application in its own right) and plugins to existing application, like VSCode (plugins are also not application in their own right).

Do you have a specific use for this as an application? Why does this need to be published seperately? Why does it need the lockfile to be _here_? (I gather `nixpkgs` is mirroring, in which case they should maintain the lock?)

---

Long version:

`package-lock.json` and its predecessor `shrinkwrap.json` are designed for usage by application creators.
Who need exact versions of dependencies to be reproducible at a given point in time for their specific app.

`unified`, `remark`, and `mdx` are libraries, not applications.
We set dependencies to version ranges, to allow downstream adopters to deduplicate and automatically receive bug and security fixes to flow from upstream packages without needing a new release from us.
`npm`, `yarn`, `pnpm` the dependency management tools for the JS ecosystem only read the `package.json` from the registry, the lockfile has no impact.
There is little to no benefit within the npm ecosystem for a _library author_ to commit the lockfile, in fact it can cause harm, hiding issues that downstream adopters would see from the test suite and CI/CD processes.

> [Nixpkgs](https://github.com/NixOS/nixpkgs) specifically relies on npm projects having lockfiles to package them.

Okay, given the context in the TL;DR am not clear why we would publish to any application registry, `nixpkgs` included.
Nor why the lockfile needs to be here in this repo.

**Comment by gekoke** [comment]
Hi, thank you for your detailed response!

> unified, remark, and mdx are libraries, not applications.
We set dependencies to version ranges, to allow downstream adopters to deduplicate and automatically receive bug and security fixes to flow from upstream packages without needing a new release from us.
npm, yarn, pnpm the dependency management tools for the JS ecosystem only read the package.json from the registry, the lockfile has no impact.
There is little to no benefit within the npm ecosystem for a library author to commit the lockfile, in fact it can cause harm, hiding issues that downstream adopters would see from the test suite and CI/CD processes.

This makes a lot of sense.

I was only looking to package the `@mdx-js/language-server` subproject, which I would consider an application, since it provides the CLI `mdx-language-server`. The lockfile needn't apply to other subprojects in the repo 
(though it seems that lockfiling subprojects separately [can't be done currently anyway](https://github.com/npm/rfcs/issues/554)).

It's likely that there are people who use editors other than VSCode and would like to consume the language server as a standalone application, and not just as part of an implementation for a VSCode plugin.

It's standard fare to request upstream to include a lockfile before packaging for nixpkgs, but I can see why it might not be appropriate in this case.

It's not a huge issue either way. Thanks for your time!

**Comment by wooorm** [comment]
afaik lock files in npm monorepos are added to the root, not to subpackages, so we can’t even add it if we wanted to.

> would like to consume the language server as a standalone application

If you *consume* it, you aren’t getting package locks.
If you are making your own application, you can choose yourself if you want package locks.

---

Closing, we don’t maintain our projects with package locks currently!

**Comment by remcohaszing** [comment]
I agree language servers classify as an application, not a library. However, its intended usage it to install it from npm, which doesn’t expose the lockfile. Whether or not we use a lockfile is an internal implementation detail you shouldn’t rely on. Also as you already pointed out, adding a lockfile would mean it’s for the entire monorepo, not just one project.

I am curious though, how would the lockfile make a difference for NixOS as a downstream user?

**Comment by gekoke** [comment]
> I am curious though, how would the lockfile make a difference for NixOS as a downstream user?

Nix does not allow network requests in packages, unless the hash of the response content is known ahead of time.
If we fetch from the npm registry, we still have to fetch the dependencies (of possibly varying versions) during build time - and to fetch the dependencies, we need to know their hashes. 

This means mirroring the npm registry [is a bunch of generated Nix code and dependency resolution](https://raw.githubusercontent.com/NixOS/nixpkgs/master/pkgs/development/node-packages/node-packages.nix).

This solution has been deemed unmaintainable. Building from the source repository using a lockfile is now preferred, if at all possible.
